### ISSUE TITLE
pebble: error when replaying WAL: pebble/record: invalid chunk

### ISSUE THREAD

**Issue by nick-jones** [cockroachdb/cockroach#68319]
**Describe the problem**

A node in one of our clusters is failing to start with the following being logged:

```
++ hostname -f
+ exec /cockroach/cockroach start --logtostderr=WARNING --certs-dir /cockroach/cockroach-certs --advertise-addr cockroachdb-1.cockroachdb.***.svc.cluster.local --http-addr 0.0.0.0 --join cockroachdb-0.cockroachdb,cockroachdb-1.cockroachdb,cockroachdb-2.cockroachdb --cache 25% --max-sql-memory 25%
Flag --logtostderr has been deprecated, use --log instead to specify 'sinks: {stderr: {filter: ...}}'.
E210802 09:27:44.365968 1 1@cli/error.go:399 ⋮ [-] 1  ‹ERROR: failed to start server: failed to create engines: pebble: error when replaying WAL: pebble/record: invalid chunk›
ERROR: failed to start server: failed to create engines: pebble: error when replaying WAL: pebble/record: invalid chunk
Failed running "start"
```

**To Reproduce**

It's unclear what triggered this. This cockroachdb cluster is running in one of our kubernetes custers. The relevant pod has been crashlooping since 30th July. We were having issues with our kube cluster on this day, and it almost certainly will have experienced at least one ungraceful shutdown. The last upgrade we ran was v21.1.5 -> v21.1.6 on 22nd July.

It's possible that some part of our infrastructure has misbehaved causing corruption of the relevant file.. if that were the case I'm interested to know the best route to recover from this.

**Expected behavior**

Node starts

**Additional data / screenshots**

Dir listing: https://gist.github.com/nick-jones/05c972baba0711eb21299960b42bf00a

**Environment:**
 - CockroachDB version: v21.1.6
 - Server OS: Linux

**Additional context**

**Comment by jbowens** [comment]
Hi @nick-jones, thanks for the report!

If you're able to provide us with any of the following files from the node, it would be helpful in diagnosing what happened:

From `/cockroach/cockroach-data/logs`:
* `cockroach-pebble.cockroachdb-1.root.2021-07-30T13_31_12Z.000067.log`
* `cockroach-pebble.cockroachdb-1.root.2021-07-27T20_23_03Z.000058.log`
* `cockroach-pebble.cockroachdb-1.root.2021-07-25T12_12_14Z.000069.log`
* `cockroach-stderr.cockroachdb-1.root.2021-07-30T12_23_52Z.000142.log`
* `cockroach.cockroachdb-1.root.2021-07-29T23_21_16Z.000058.log`
* `cockroach.cockroachdb-1.root.2021-07-30T12_35_25Z.000195.log`
* `cockroach.cockroachdb-1.root.2021-07-30T13_30_39Z.000067.log`
* `cockroach.cockroachdb-1.root.2021-07-30T13_31_28Z.000118.log`
* `cockroach.cockroachdb-1.root.2021-07-30T13_31_37Z.000145.log`

From `/cockroach/cockroach-data/`:
* `CURRENT`
* `MANIFEST-150756`
* `MANIFEST-155269`
* `155268.log`
* `155252.log`
* `155231.log`
* `155216.log`
* `155199.log`
* `155184.log`
* `155163.log`

If you can provide these files, you can email them to me at jackson (at cockroachlabs.com). At first glance it does look like an ungraceful shutdown exercised a bug.

**Comment by jbowens** [comment]
Hi @nick-jones, thanks again for the report. I verified that this is a bug.

When the storage engine opens, it replays recent write-ahead logs. When the storage engine replays the write-ahead logs, it expects every write-ahead log except for the most recent one to be cleanly closed with a 'EOF' trailer (cockroachdb/pebble#864). If any of the older write-ahead logs don't end cleanly, we consider it corruption.

After the storage engine has replayed all its write-ahead logs, it creates a new write-ahead log that it will use for future writes.

On the 31st your node terminated and started up again, presumably from the k8s troubles. The most recent write-ahead log `155252.log` was the only one with an unclean tail, as expected. The storage engine replayed all the write-ahead logs, _created a new one_ `155268.log` and terminated immediately afterwards. On all subsequent node stats, the storage engine no longer considered `155252.log` to be the most-recent write-ahead log and crashed when it observed its unclean tail.

I have a pull request (cockroachdb/pebble#1212) to fix this by ensuring we mark the old write-ahead logs as replayed before we create the new write-ahead log file.

For now the safest action is to just decommission the node and replace it. Let me know if you for some reason you can't and need to recover the node.

**Comment by nick-jones** [comment]
> Hi @nick-jones, thanks again for the report. I verified that this is a bug.
> 
> When the storage engine opens, it replays recent write-ahead logs. When the storage engine replays the write-ahead logs, it expects every write-ahead log except for the most recent one to be cleanly closed with a 'EOF' trailer ([cockroachdb/pebble#864](https://github.com/cockroachdb/pebble/issues/864)). If any of the older write-ahead logs don't end cleanly, we consider it corruption.
> 
> After the storage engine has replayed all its write-ahead logs, it creates a new write-ahead log that it will use for future writes.
> 
> On the 31st your node terminated and started up again, presumably from the k8s troubles. The most recent write-ahead log `155252.log` was the only one with an unclean tail, as expected. The storage engine replayed all the write-ahead logs, _created a new one_ `155268.log` and terminated immediately afterwards. On all subsequent node stats, the storage engine no longer considered `155252.log` to be the most-recent write-ahead log and crashed when it observed its unclean tail.
> 
> I have a pull request ([cockroachdb/pebble#1212](https://github.com/cockroachdb/pebble/pull/1212)) to fix this by ensuring we mark the old write-ahead logs as replayed before we create the new write-ahead log file.

Nice one figuring that out so quickly :+1:

> 
> For now the safest action is to just decommission the node and replace it. Let me know if you for some reason you can't and need to recover the node.

We've successfully decommissioned the node and replaced it. The cluster appears to be happy and healthy again, so it looks like we're all good. Thanks for your help with this.

**Comment by nick-jones** [comment]
@jbowens we noticed we have another node (different cluster) failing with a similar but not identical error message: 

```
++ hostname -f
+ exec /cockroach/cockroach start --logtostderr=WARNING --certs-dir /cockroach/cockroach-certs --advertise-addr cockroachdb-1.cockroachdb.***.svc.cluster.local --http-addr 0.0.0.0 --join cockroachdb-0.cockroachdb,cockroachdb-1.cockroachdb,cockroachdb-2.cockroachdb --cache 25% --max-sql-memory 25%
Flag --logtostderr has been deprecated, use --log instead to specify 'sinks: {stderr: {filter: ...}}'.
E210809 16:52:43.193293 1 1@cli/error.go:399 ⋮ [-] 1  ‹ERROR: connection lost.›
E210809 16:52:43.193293 1 1@cli/error.go:399 ⋮ [-] 1 +‹failed to start server: failed to create engines: pebble: error when replaying WAL: unexpected EOF›
ERROR: connection lost.

failed to start server: failed to create engines: pebble: error when replaying WAL: unexpected EOF
Failed running "start"
```

Is this likely to be the same problem? Happy to raise a separate issue if not.

**Comment by nick-jones** [comment]
> Hey @nick-jones, I just checked the code and yeah, it's likely to be the same problem. Just to make sure, in this most recent instance the process restarted multiple times in rapid succession?

With this instance we first started seeing the error logged on the 4th August. I'm not really sure what constitutes rapid here, but this is what I see from grepping the logs for CockroachDB starting:

```
sh-4.4# grep 'cli/start.go.*CockroachDB CCL v21.1.6' cockroach-data/logs/cockroach.cockroachdb-1.root.2021-08-04T* | awk '{print $2}'
08:53:09.102408
08:53:51.908659
08:54:35.160158
08:58:24.542415
09:00:12.858926
09:00:58.093180
09:03:15.522382
09:06:48.234119
09:12:35.424027
09:13:18.163400
09:19:05.380356
09:19:48.731580
09:25:39.753741
09:26:25.143956
09:34:59.568453
09:36:19.095497
09:42:52.466910
09:43:39.654968
09:48:45.167100
09:54:44.857266
09:55:30.918725
10:01:21.899372
10:06:45.990356
10:12:09.867908
10:17:21.759803
10:22:44.545523
<snip>
```

The first log entry we have for the error is:

```
sh-4.4# grep 'error when replaying WAL: unexpected EOF' cockroach-data/logs/cockroach.cockroachdb-1.root.2021-08-0*  | head -n1
cockroach-data/logs/cockroach.cockroachdb-1.root.2021-08-04T10_01_21Z.001062.log:E210804 10:01:39.877141 1 1@cli/error.go:399 ⋮ [-] 9 +‹failed to start server: failed to create engines: pebble: error when replaying WAL: unexpected EOF›
```

Tail of the logs previous to this:

```
sh-4.4# tail -n 5 cockroach-data/logs/cockroach.cockroachdb-1.root.2021-08-04T09_55_30Z.001013.log 
I210804 09:55:30.940374 1 1@cli/start.go:1129 ⋮ [-] 7  GEOS loaded from directory ‹/usr/local/lib/cockroach›
I210804 09:55:30.940491 1 1@cli/start.go:528 ⋮ [-] 8  starting cockroach node
I210804 09:56:12.208365 1 1@cli/start.go:755 ⋮ [-] 9  received signal 'terminated'
I210804 09:56:12.208829 1 1@cli/start.go:839 ⋮ [-] 10  initiating graceful shutdown of server
I210804 09:56:12.209341 1 1@cli/start.go:896 ⋮ [-] 11  too early to drain; used hard shutdown instead
sh-4.4# tail -n 5 cockroach-data/logs/cockroach.cockroachdb-1.root.2021-08-04T09_54_44Z.000965.log 
I210804 09:54:44.867003 1 1@cli/start.go:1129 ⋮ [-] 7  GEOS loaded from directory ‹/usr/local/lib/cockroach›
I210804 09:54:44.867059 1 1@cli/start.go:528 ⋮ [-] 8  starting cockroach node
I210804 09:55:27.385955 1 1@cli/start.go:755 ⋮ [-] 9  received signal 'terminated'
I210804 09:55:27.386155 1 1@cli/start.go:839 ⋮ [-] 10  initiating graceful shutdown of server
I210804 09:55:27.386294 1 1@cli/start.go:896 ⋮ [-] 11  too early to drain; used hard shutdown instead
sh-4.4# tail -n 5 cockroach-data/logs/cockroach.cockroachdb-1.root.2021-08-04T09_48_45Z.000920.log 
I210804 09:48:45.635909 1 1@cli/start.go:1129 ⋮ [-] 7  GEOS loaded from directory ‹/usr/local/lib/cockroach›
I210804 09:48:45.635980 1 1@cli/start.go:528 ⋮ [-] 8  starting cockroach node
I210804 09:49:27.419001 1 1@cli/start.go:755 ⋮ [-] 9  received signal 'terminated'
I210804 09:49:27.419193 1 1@cli/start.go:839 ⋮ [-] 10  initiating graceful shutdown of server
I210804 09:49:27.419284 1 1@cli/start.go:896 ⋮ [-] 11  too early to drain; used hard shutdown instead
sh-4.4# tail -n 5 cockroach-data/logs/cockroach.cockroachdb-1.root.2021-08-04T09_43_39Z.000867.log 
W210804 09:48:09.251870 2445 kv/kvserver/store_raft.go:533 ⋮ [n2,s2,r14/2:‹/Table/1{8-9}›] 187  handle raft ready: 44.7s [applied=3, batches=2, state_assertions=0]
W210804 09:48:09.258958 2500 kv/kvserver/store_raft.go:533 ⋮ [n2,s2,r278/2:‹/System/tsd/cr.node.txn.re…›] 188  handle raft ready: 30.0s [applied=36, batches=1, state_assertions=0]
W210804 09:48:09.243620 2457 kv/kvserver/store_raft.go:533 ⋮ [n2,s2,r252/2:‹/System/tsd/cr.node.{cl…-di…}›] 171  handle raft ready: 70.1s [applied=82, batches=1, state_assertions=0]
W210804 09:48:12.824143 4608 server/node.go:768 ⋮ [n2,summaries] 189  health alerts detected: ‹{Alerts:[{StoreID:2 Category:METRICS Description:requests.slow.raft Value:1}]}›
I210804 09:48:16.253832 2760 2@server/status/runtime.go:569 ⋮ [n2] 190  runtime stats: 1.1 GiB RSS, 385 goroutines (stacks: 7.1 MiB), 649 MiB/712 MiB Go alloc/total (heap fragmentation: 16 MiB, heap reserved: 656 KiB, heap released: 32 MiB), 233 MiB/266 MiB CGO alloc/total (108.8 CGO/sec), 10.8/3.1 %(u/s)time, 0.0 %gc (0x), 9.2 MiB/203 KiB (r/w)net
```

So 09:48:16 is the last time we saw it properly running.

**Comment by jbowens** [comment]
Thanks @nick-jones — yeah, I think it is the same issue. The process restarts aren't as rapid as I expected, but these log lines indicate that something was slowing the process to a crawl:

```
W210804 09:48:09.251870 2445 kv/kvserver/store_raft.go:533 ⋮ [n2,s2,r14/2:‹/Table/1{8-9}›] 187  handle raft ready: 44.7s [applied=3, batches=2, state_assertions=0]
W210804 09:48:09.258958 2500 kv/kvserver/store_raft.go:533 ⋮ [n2,s2,r278/2:‹/System/tsd/cr.node.txn.re…›] 188  handle raft ready: 30.0s [applied=36, batches=1, state_assertions=0]
W210804 09:48:09.243620 2457 kv/kvserver/store_raft.go:533 ⋮ [n2,s2,r252/2:‹/System/tsd/cr.node.{cl…-di…}›] 171  handle raft ready: 70.1s [applied=82, batches=1, state_assertions=0]
W210804 09:48:12.824143 4608 server/node.go:768 ⋮ [n2,summaries] 189  health alerts detected: ‹{Alerts:[{StoreID:2 Category:METRICS Description:requests.slow.raft Value:1}]}›
```
Was there an issue with or modification of an elastic storage volume? In any case, based on those raft ready latencies of 30.0s, 44.7s and 70.1s, the node definitely could've still been replaying WALs at the time one of the later processes terminated, exercising that same bug.

**Comment by nick-jones** [comment]
> Thanks @nick-jones — yeah, I think it is the same issue.

Thanks for confirming :+1:

> The process restarts aren't as rapid as I expected, but these log lines indicate that something was slowing the process to a crawl:
> 
> ```
> W210804 09:48:09.251870 2445 kv/kvserver/store_raft.go:533 ⋮ [n2,s2,r14/2:‹/Table/1{8-9}›] 187  handle raft ready: 44.7s [applied=3, batches=2, state_assertions=0]
> W210804 09:48:09.258958 2500 kv/kvserver/store_raft.go:533 ⋮ [n2,s2,r278/2:‹/System/tsd/cr.node.txn.re…›] 188  handle raft ready: 30.0s [applied=36, batches=1, state_assertions=0]
> W210804 09:48:09.243620 2457 kv/kvserver/store_raft.go:533 ⋮ [n2,s2,r252/2:‹/System/tsd/cr.node.{cl…-di…}›] 171  handle raft ready: 70.1s [applied=82, batches=1, state_assertions=0]
> W210804 09:48:12.824143 4608 server/node.go:768 ⋮ [n2,summaries] 189  health alerts detected: ‹{Alerts:[{StoreID:2 Category:METRICS Description:requests.slow.raft Value:1}]}›
> ```
> 
> Was there an issue with or modification of an elastic storage volume? In any case, based on those raft ready latencies of 30.0s, 44.7s and 70.1s, the node definitely could've still been replaying WALs at the time one of the later processes terminated, exercising that same bug.

It looks like there were issues with k8s node that this particular cockroachdb node was scheduled on at the time.. so much so we have a gap in our metrics during that period:

<img width="806" alt="Screenshot 2021-08-13 at 18 15 08" src="https://user-images.githubusercontent.com/350792/129396331-68577b44-e67f-4c42-8cd7-d10752ec0817.png">

So yeah, I think it's fair to assume things weren't in a healthy place.

**Comment by jbowens** [comment]
The fix was backported to 21.1 today in #68897, so it should go out with 21.1.8.

I'm going to close this out, but feel free to still ping me if you want to double check other occurrences.
